tests/core/network/test_enhanced_swarm.py

```diff
@@ -0,0 +1,416 @@
+    """Mock connection for testing."""
+
+    def __init__(self, peer_id: ID, is_closed: bool = False):
+        self.peer_id = peer_id
+        self._is_closed = is_closed
+        self.stream_count = 0
+        # Mock the muxed_conn attribute that Swarm expects
+        self.muxed_conn = Mock()
+        self.muxed_conn.peer_id = peer_id
+
+    async def close(self):
+        self._is_closed = True
+
+    @property
+    def is_closed(self) -> bool:
+        return self._is_closed
+
+    async def new_stream(self) -> INetStream:
+        self.stream_count += 1
+        return Mock(spec=INetStream)
+
+    def get_streams(self) -> tuple[INetStream, ...]:
+        """Mock implementation of get_streams."""
+        return tuple()
+
+    def get_transport_addresses(self) -> list[Multiaddr]:
+        """Mock implementation of get_transport_addresses."""
+        return []
+
+
+class MockNetStream(INetStream):
+    """Mock network stream for testing."""
+
+    def __init__(self, peer_id: ID):
+        self.peer_id = peer_id
+
+
+@pytest.mark.trio
+async def test_retry_config_defaults():
+    """Test RetryConfig default values."""
+    config = RetryConfig()
+    assert config.max_retries == 3
+    assert config.initial_delay == 0.1
+    assert config.max_delay == 30.0
+    assert config.backoff_multiplier == 2.0
+    assert config.jitter_factor == 0.1
+
+
+@pytest.mark.trio
+async def test_connection_config_defaults():
+    """Test ConnectionConfig default values."""
+    config = ConnectionConfig()
+    assert config.max_connections_per_peer == 3
+    assert config.connection_timeout == 30.0
+    assert config.enable_connection_pool is True
+    assert config.load_balancing_strategy == "round_robin"
+
+
+@pytest.mark.trio
+async def test_connection_pool_basic_operations():
+    """Test basic ConnectionPool operations."""
+    pool = ConnectionPool(max_connections_per_peer=2)
+    peer_id = ID(b"QmTest")
+
+    # Test empty pool
+    assert not pool.has_connection(peer_id)
+    assert pool.get_connection(peer_id) is None
+
+    # Add connection
+    conn1 = MockConnection(peer_id)
+    pool.add_connection(peer_id, conn1, "addr1")
+    assert pool.has_connection(peer_id)
+    assert pool.get_connection(peer_id) == conn1
+
+    # Add second connection
+    conn2 = MockConnection(peer_id)
+    pool.add_connection(peer_id, conn2, "addr2")
+    assert len(pool.peer_connections[peer_id]) == 2
+
+    # Test round-robin - should cycle through connections
+    first_conn = pool.get_connection(peer_id, "round_robin")
+    second_conn = pool.get_connection(peer_id, "round_robin")
+    third_conn = pool.get_connection(peer_id, "round_robin")
+
+    # Should cycle through both connections
+    assert first_conn in [conn1, conn2]
+    assert second_conn in [conn1, conn2]
+    assert third_conn in [conn1, conn2]
+    assert first_conn != second_conn or second_conn != third_conn
+
+    # Test least loaded - set different stream counts
+    conn1.stream_count = 5
```

Comment:
Is `stream_count` managed in the pool module so that `inc` `dec` are updated when streams opens / closes ?